File: platform/pylib/scout_common/config/validator.py
"""Validation utilities for configuration values."""

import re
from typing import Any, TypeVar


from .errors import (
    missing_required,
    invalid_value,
    invalid_format,
    out_of_range,
    invalid_choice,
)
from .parser import parse_url


T = TypeVar("T", int, float)


def validate_required(key: str, value: str) -> None:
    """
    Check if a value is non-empty.
    
    Args:
        key: Configuration key name
        value: Value to validate
        
    Raises:
        Error: If value is empty
    """
    if not value or not value.strip():
        raise missing_required(key)


def validate_range(key: str, value: T, min_val: T, max_val: T) -> None:
    """
    Check if a numeric value is within the specified range (inclusive).
    
    Args:
        key: Configuration key name
        value: Value to validate
        min_val: Minimum allowed value (inclusive)
        max_val: Maximum allowed value (inclusive)
        
    Raises:
        Error: If value is outside the range
    """
    if value < min_val or value > max_val:
        raise out_of_range(key, value, min_val, max_val)


def validate_min_max(
    key: str,
    value: T,
    min_val: T | None = None,
    max_val: T | None = None,
) -> None:
    """
    Check if a numeric value satisfies min/max constraints.
    
    Pass None for min_val or max_val to skip that check.
    
    Args:
        key: Configuration key name
        value: Value to validate
        min_val: Minimum allowed value (optional)
        max_val: Maximum allowed value (optional)
        
    Raises:
        Error: If value violates constraints
    """
    if min_val is not None and value < min_val:
        raise out_of_range(key, value, min_val, "infinity")
    if max_val is not None and value > max_val:
        raise out_of_range(key, value, "infinity", max_val)


def validate_choice(key: str, value: str, allowed_values: list[str]) -> None:
    """
    Check if a value is in the allowed set.
    
    Args:
        key: Configuration key name
        value: Value to validate
        allowed_values: List of allowed values
        
    Raises:
        Error: If value is not in allowed set
    """
    if value not in allowed_values:
        raise invalid_choice(key, value, allowed_values)


def validate_pattern(key: str, value: str, pattern: str) -> None:
    """
    Check if a value matches the given regex pattern.
    
    Args:
        key: Configuration key name
        value: Value to validate
        pattern: Regex pattern
        
    Raises:
        Error: If value doesn't match pattern
    """
    if not re.match(pattern, value):
        raise invalid_format(key, value, f"pattern: {pattern}")


def validate_url(key: str, value: str) -> None:
    """
    Check if a value is a valid URL with required scheme.
    
    Args:
        key: Configuration key name
        value: URL string to validate
        
    Raises:
        Error: If URL is invalid
    """
    try:
        parse_url(value)
    except ValueError as e:
        raise invalid_format(key, value, "valid URL with scheme") from e


def validate_port(key: str, port: int) -> None:
    """
    Check if a port number is in the valid range (1-65535).
    
    Args:
        key: Configuration key name
        port: Port number to validate
        
    Raises:
        Error: If port is outside valid range
    """
    validate_range(key, port, 1, 65535)


def validate_non_zero(key: str, value: T) -> None:
    """
    Check if a numeric value is non-zero.
    
    Args:
        key: Configuration key name
        value: Value to validate
        
    Raises:
        Error: If value is zero
    """
    if value == 0:
        raise invalid_value(key, "0", "value must be non-zero")


def validate_positive(key: str, value: T) -> None:
    """
    Check if a numeric value is positive (> 0).
    
    Args:
        key: Configuration key name
        value: Value to validate
        
    Raises:
        Error: If value is not positive
    """
    if value <= 0:
        raise invalid_value(key, str(value), "value must be positive")


def validate_non_negative(key: str, value: T) -> None:
    """
    Check if a numeric value is non-negative (>= 0).
    
    Args:
        key: Configuration key name
        value: Value to validate
        
    Raises:
        Error: If value is negative
    """
    if value < 0:
        raise invalid_value(key, str(value), "value must be non-negative")


def validate_min_length(key: str, value: str, min_length: int) -> None:
    """
    Check if a string meets minimum length requirement.
    
    Args:
        key: Configuration key name
        value: String to validate
        min_length: Minimum required length
        
    Raises:
        Error: If string is too short
    """
    if len(value) < min_length:
        raise invalid_value(
            key,
            value,
            f"must be at least {min_length} characters",
        )


def validate_max_length(key: str, value: str, max_length: int) -> None:
    """
    Check if a string does not exceed maximum length.
    
    Args:
        key: Configuration key name
        value: String to validate
        max_length: Maximum allowed length
        
    Raises:
        Error: If string is too long
    """
    if len(value) > max_length:
        raise invalid_value(
            key,
            value,
            f"must be at most {max_length} characters",
        )

File: platform/pkg/config/validator.go
// platform/pkg/config/validator.go
package config

import (
	"fmt"
	"regexp"
)

// ValidateRequired checks if a value is non-empty.
func ValidateRequired(key, value string) error {
	if value == "" {
		return MissingRequired(key)
	}
	return nil
}

// ValidateRange checks if a numeric value is within the specified range (inclusive).
func ValidateRange[T int | int64 | float64](key string, value, min, max T) error {
	if value < min || value > max {
		return OutOfRange(key, value, min, max)
	}
	return nil
}

// ValidateMinMax checks if a numeric value satisfies min/max constraints.
// Pass nil for min or max to skip that check.
func ValidateMinMax[T int | int64 | float64](key string, value T, min, max *T) error {
	if min != nil && value < *min {
		return OutOfRange(key, value, *min, "infinity")
	}
	if max != nil && value > *max {
		return OutOfRange(key, value, "infinity", *max)
	}
	return nil
}

// ValidateChoice checks if a value is in the allowed set.
func ValidateChoice(key, value string, allowedValues []string) error {
	for _, allowed := range allowedValues {
		if value == allowed {
			return nil
		}
	}
	return InvalidChoice(key, value, allowedValues)
}

// ValidatePattern checks if a value matches the given regex pattern.
func ValidatePattern(key, value, pattern string) error {
	matched, err := regexp.MatchString(pattern, value)
	if err != nil {
		return InvalidFormat(key, value, fmt.Sprintf("regex pattern: %s", pattern))
	}
	if !matched {
		return InvalidFormat(key, value, fmt.Sprintf("pattern: %s", pattern))
	}
	return nil
}

// ValidateURL checks if a value is a valid URL with required scheme.
func ValidateURL(key, value string) error {
	_, err := ParseURL(value)
	if err != nil {
		return InvalidFormat(key, value, "valid URL with scheme")
	}
	return nil
}

// ValidatePort checks if a port number is in the valid range (1-65535).
func ValidatePort(key string, port int) error {
	return ValidateRange(key, port, 1, 65535)
}

// ValidateNonZero checks if a numeric value is non-zero.
func ValidateNonZero[T int | int64 | float64](key string, value T) error {
	if value == 0 {
		return InvalidValue(key, "0", "value must be non-zero")
	}
	return nil
}

// ValidatePositive checks if a numeric value is positive (> 0).
func ValidatePositive[T int | int64 | float64](key string, value T) error {
	if value <= 0 {
		return InvalidValue(key, fmt.Sprintf("%v", value), "value must be positive")
	}
	return nil
}

// ValidateNonNegative checks if a numeric value is non-negative (>= 0).
func ValidateNonNegative[T int | int64 | float64](key string, value T) error {
	if value < 0 {
		return InvalidValue(key, fmt.Sprintf("%v", value), "value must be non-negative")
	}
	return nil
}

// ValidateMinLength checks if a string meets minimum length requirement.
func ValidateMinLength(key, value string, minLength int) error {
	if len(value) < minLength {
		return InvalidValue(
			key,
			value,
			fmt.Sprintf("must be at least %d characters", minLength),
		)
	}
	return nil
}

// ValidateMaxLength checks if a string does not exceed maximum length.
func ValidateMaxLength(key, value string, maxLength int) error {
	if len(value) > maxLength {
		return InvalidValue(
			key,
			value,
			fmt.Sprintf("must be at most %d characters", maxLength),
		)
	}
	return nil
}